lib/Partitioner/Partitioner.cpp

```diff
@@ -492,7 +501,78 @@ NodeToFunctionMap Partitioner::selectPartitions(Function *F,
 /// the number of DAGNodes) is larger than the number of devices. E.g:
 /// node1(6GB) -> node2(14GB) -> node3(6GB). The memory limitation is 16GB, and
 /// there is only 2 devices.
-void Partitioner::adjustLogicalDeviceID() {}
+llvm::Error Partitioner::adjustLogicalDeviceID(NodeToFunctionMap &mapping) {
+  for (auto &dag : partitions_) {
+    // Traver the DAGNodes, and get list of with each backendKind.
+    std::map<BackendKind, std::vector<DAGNode *>> backendKindNum;
+    for (auto &node : dag.nodes) {
+      auto backendKind = node->backendKind;
+      if (backendKindNum.find(backendKind) == backendKindNum.end()) {
+        backendKindNum.emplace(
+            std::make_pair(backendKind, std::vector<DAGNode *>{node.get()}));
+      } else {
+        backendKindNum[backendKind].push_back(node.get());
+      }
+    }
+
+    // For each type of the backend, find out the number of partitions, and the
+    // number of avaiable devices with this backend. If the former is larger
+    // than the later, we will try to assign several partitions into 1 device by
+    // assigning those partitions with the same logicalDevice ID. If there is no
+    // way to make the required devices number fewer or equal to the actual
+    // number of devices, we just skip this adjustment.
+    for (const auto &p : backendKindNum) {
+      if (p.second.size() <= backendMap_[p.first].num) {
+        // There is enough device with this backendKind, no need to adjust the
+        // logical ID.
+        continue;
+      }
+      // Get the list of DAGNode with current BackendKind, and sort it based on
+      // used memory from min to max.
+      std::vector<std::pair<DAGNode *, uint64_t>> nodeSize;
+      for (size_t i = 0, e = p.second.size(); i < e; i++) {
+        Function *function = module_->getFunction(p.second[i]->name);
+        uint64_t totalMem = mapping.getGraphMemInfo(function).getTotalMemSize();
+        nodeSize.push_back(std::make_pair(p.second[i], totalMem));
+      }
+      std::sort(nodeSize.begin(), nodeSize.end(), sortMinMemory);
+
+      // Assume we have n devices and m partitions, where m > n. If these m
+      // partitions can be assigned to n devices, there must be 1 device have at
+      // least (m - n + 1) partitions. Therefore, we just need to check the
```

Comment:
> If these m partitions can be assigned to n devices, there must be 1 device have at least (m - n + 1) partitions. 

Is that true? E.g. if we have 10 partitions and 5 devices, you say that at least one device should get (10 - 5 + 1) = 6 partitions. Why? Why can't each device e.g. get two partitions?

Reply:
Good catch!  Thanks for pointing this! I thought the putting n+1 items to n boxes issue... I will rewrite this algorithm.